packages/config-loader/src/lib/schema/collect.ts

```diff
@@ -0,0 +1,180 @@
+
+type Item = {
+  name: string;
+  parentPath?: string;
+};
+
+const req =
+  typeof __non_webpack_require__ === 'undefined'
+    ? require
+    : __non_webpack_require__;
+
+/**
+ * This collects all known config schemas across all dependencies of the app.
+ */
+export async function collectConfigSchemas(
+  packageNames: string[],
+): Promise<ConfigSchemaPackageEntry[]> {
+  const visitedPackages = new Set<string>();
+  const schemas = Array<ConfigSchemaPackageEntry>();
+  const tsSchemaPaths = Array<string>();
+  const currentDir = await fs.realpath(process.cwd());
+
+  async function processItem({ name, parentPath }: Item) {
+    // Ensures that we only process each package once. We don't bother with
+    // loading in schemas from duplicates of different versions, as that's not
+    // supported by Backstage right now anyway. We may want to change that in
+    // the future though, if it for example becomes possible to load in two
+    // different versions of e.g. @backstage/core at once.
+    if (visitedPackages.has(name)) {
+      return;
+    }
+    visitedPackages.add(name);
+
+    let pkgPath: string;
+    try {
+      pkgPath = req.resolve(
+        `${name}/package.json`,
+        parentPath && {
+          paths: [parentPath],
+        },
+      );
+    } catch {
+      // We can somewhat safely ignore packages that don't export package.json,
+      // as they are likely not part of the Backstage ecosystem anyway.
+      return;
+    }
+
+    const pkg = await fs.readJson(pkgPath);
+    const depNames = [
+      ...Object.keys(pkg.dependencies ?? {}),
+      ...Object.keys(pkg.peerDependencies ?? {}),
+    ];
+
+    // TODO(Rugvip): Trying this out to avoid having to traverse the full dependency graph,
+    //               since that's pretty slow. We probably need a better way to determine when
+    //               we've left the Backstage ecosystem, but this will do for now.
+    const hasSchema = 'configSchema' in pkg;
+    const hasBackstageDep = depNames.some(_ => _.startsWith('@backstage/'));
+    if (!hasSchema && !hasBackstageDep) {
+      return;
+    }
+    if (hasSchema) {
+      if (typeof pkg.configSchema === 'string') {
+        const isJson = pkg.configSchema.endsWith('.json');
+        const isDts = pkg.configSchema.endsWith('.d.ts');
+        if (!isJson && !isDts) {
+          throw new Error(
+            `Config schema files must be .json or .d.ts, got ${pkg.configSchema}`,
+          );
+        }
+        if (isDts) {
+          tsSchemaPaths.push(
+            relativePath(
+              currentDir,
+              resolvePath(dirname(pkgPath), pkg.configSchema),
+            ),
+          );
+        } else {
+          const path = resolvePath(dirname(pkgPath), pkg.configSchema);
+          const value = await fs.readJson(path);
+          schemas.push({
+            value,
+            path: relativePath(currentDir, path),
+          });
+        }
+      } else {
```

Comment:
I'm kinda wondering if we should instead always make this a path, and always make it `.d.ts`, and always generate one (but empty) with every plugin, always have it present in `files`. More opinionated, fewer choices, easier to get going, fewer steps to forget (like "oh i have to add it to the files section too? i didn't realise")

Reply:
Yeah if this ends up being something almost every plugin does then I think we could add that in the future. Still hoping that adding config isn't something that every plugin does or something we should encourage :p